src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,686 @@
+        // Store the updated claimed amount
+        withdrawsClaimed += request.assets;
+
+        assets = request.assets;
+
+        emit RedeemClaimed(msg.sender, requestId, assets);
+
+        // transfer the liquidity asset to the withdrawer
+        IERC20(liquidityAsset).transfer(msg.sender, assets);
+    }
+
+    /// @dev Updates the claimable amount in the ARM's withdrawal queue.
+    /// That's the amount that is used to check if a request can be claimed or not.
+    function _updateWithdrawalQueueLiquidity() internal {
+        // Load the claimable amount from storage into memory
+        uint256 withdrawsClaimableMem = withdrawsClaimable;
+
+        // Check if the claimable amount is less than the queued amount
+        uint256 queueShortfall = withdrawsQueued - withdrawsClaimableMem;
+
+        // No need to do anything is the withdrawal queue is fully funded
+        if (queueShortfall == 0) {
+            return;
+        }
+
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // Of the claimable withdrawal requests, how much is unclaimed?
+        // That is, the amount of the liquidity assets that is currently allocated for the withdrawal queue
+        uint256 allocatedLiquidity = withdrawsClaimableMem - withdrawsClaimed;
+
+        // If there is no unallocated liquidity assets then there is nothing to add to the queue
+        if (liquidityBalance <= allocatedLiquidity) {
+            return;
+        }
+
+        uint256 unallocatedLiquidity = liquidityBalance - allocatedLiquidity;
+
+        // the new claimable amount is the smaller of the queue shortfall or unallocated weth
+        uint256 addedClaimable = queueShortfall < unallocatedLiquidity ? queueShortfall : unallocatedLiquidity;
+
+        // Store the new claimable amount back to storage
+        withdrawsClaimable = SafeCast.toUint128(withdrawsClaimableMem + addedClaimable);
+    }
+
+    /// @dev Calculate how much of the liquidity asset in the ARM is not reserved for the withdrawal queue.
+    // That is, it is available to be swapped.
+    function _liquidityAvailable() internal view returns (uint256) {
+        // The amount of WETH that is still to be claimed in the withdrawal queue
+        uint256 outstandingWithdrawals = withdrawsQueued - withdrawsClaimed;
+
+        // The amount of the liquidity asset is in the ARM
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // If there is not enough liquidity assets in the ARM to cover the outstanding withdrawals
+        if (liquidityBalance <= outstandingWithdrawals) {
+            return 0;
+        }
+
+        return liquidityBalance - outstandingWithdrawals;
+    }
+
+    /// @notice The total amount of assets in the ARM and external withdrawal queue,
+    /// less the liquidity assets reserved for the ARM's withdrawal queue and accrued fees.
+    function totalAssets() public view virtual returns (uint256) {
```

Comment:
🌑 So here's an attack route I haven't worked out fully yet.

We plan to allow trades that will be under 1:1. These trades will make the number (weth+stETH+queue) go down. Would it be possible to engineer a situation where (weth+stETH+queue) - (exits + old fees + new fees) gets to be some kind of low number that allows for fun rounding shenanigans.

Reply:
So if the (weth+stETH+queue) was going down, we could be redeeming more per LP token.